File: go/tests/anomaly_signal_test.go
// anomaly_signal_test.go
//
// Validates NewAnomalySignal — confidence is clamped to [0.0, 1.0], evidence
// defaults to an empty map when nil, and stamped fields (ID, DetectedAt) are
// populated. Clamp vectors come from fixtures/anomaly_signal_schema.json.
// Float comparisons use epsilon = 1e-6 (clamp is a no-loss operation).

package circleai_test

import (
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

// ---------------------------------------------------------------------------
// Fixture types
// ---------------------------------------------------------------------------

type anomalyFixture struct {
	ThreatVector  threatVectorBlock `json:"threatVector"`
	ClampVectors  []clampVector     `json:"clamp_vectors"`
}

type threatVectorBlock struct {
	Values []threatVectorEntry `json:"values"`
}

type threatVectorEntry struct {
	Name    string `json:"name"`
	Ordinal int    `json:"ordinal"`
}

type clampVector struct {
	ID                  string  `json:"id"`
	InputConfidence     float64 `json:"input_confidence"`
	ExpectedConfidence  float64 `json:"expected_confidence"`
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func loadAnomalyFixture(t *testing.T) anomalyFixture {
	t.Helper()
	path := filepath.Join(fixturesDir(t), "anomaly_signal_schema.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read anomaly_signal_schema.json: %v", err)
	}
	var fix anomalyFixture
	if err := json.Unmarshal(data, &fix); err != nil {
		t.Fatalf("failed to parse anomaly_signal_schema.json: %v", err)
	}
	return fix
}

// ---------------------------------------------------------------------------
// ThreatVector ordinals — must stay stable across language ports
// ---------------------------------------------------------------------------

func TestThreatVector_OrdinalsMatchFixture(t *testing.T) {
	fix := loadAnomalyFixture(t)

	if len(fix.ThreatVector.Values) == 0 {
		t.Fatal("no threatVector values in fixture")
	}

	want := map[string]circleai.ThreatVector{
		"MemoryAnomaly":         circleai.ThreatVectorMemoryAnomaly,
		"ControlFlowDrift":      circleai.ThreatVectorControlFlowDrift,
		"PrivilegeEscalation":   circleai.ThreatVectorPrivilegeEscalation,
		"BiometricSpoofAttempt": circleai.ThreatVectorBiometricSpoofAttempt,
		"NetworkPivot":          circleai.ThreatVectorNetworkPivot,
		"StateCorruption":       circleai.ThreatVectorStateCorruption,
		"AgentPatchRejected":    circleai.ThreatVectorAgentPatchRejected,
		"Unknown":               circleai.ThreatVectorUnknown,
	}

	for _, entry := range fix.ThreatVector.Values {
		got, ok := want[entry.Name]
		if !ok {
			t.Errorf("fixture has unknown ThreatVector name %q", entry.Name)
			continue
		}
		if int(got) != entry.Ordinal {
			t.Errorf("ThreatVector%s: got ordinal %d, want %d", entry.Name, int(got), entry.Ordinal)
		}
	}

	if len(fix.ThreatVector.Values) != len(want) {
		t.Errorf("fixture has %d ThreatVector entries, Go port has %d",
			len(fix.ThreatVector.Values), len(want))
	}
}

// ---------------------------------------------------------------------------
// Confidence clamp — table-driven against fixture
// ---------------------------------------------------------------------------

func TestAnomalySignal_ConfidenceClamp(t *testing.T) {
	const epsilon = 1e-6

	fix := loadAnomalyFixture(t)

	if len(fix.ClampVectors) == 0 {
		t.Fatal("no clamp_vectors in fixture")
	}

	for _, v := range fix.ClampVectors {
		v := v
		t.Run(v.ID, func(t *testing.T) {
			sig := circleai.NewAnomalySignal(
				circleai.ThreatVectorMemoryAnomaly,
				float32(v.InputConfidence),
				"Circle.AI.Test",
				"clamp vector "+v.ID,
				nil,
			)
			if sig == nil {
				t.Fatal("NewAnomalySignal returned nil")
			}
			got := float64(sig.Confidence)
			if math.Abs(got-v.ExpectedConfidence) > epsilon {
				t.Errorf("[%s] confidence: input=%v got=%v want=%v",
					v.ID, v.InputConfidence, got, v.ExpectedConfidence)
			}
		})
	}
}

// ---------------------------------------------------------------------------
// Factory behaviour — ID, DetectedAt, evidence default
// ---------------------------------------------------------------------------

func TestNewAnomalySignal_StampsIDAndTimestamp(t *testing.T) {
	before := time.Now().UTC()
	sig := circleai.NewAnomalySignal(
		circleai.ThreatVectorBiometricSpoofAttempt,
		0.42,
		"Circle.AI.Identity",
		"liveness check failed",
		map[string]string{"reason": "no_blink"},
	)
	after := time.Now().UTC()

	if sig == nil {
		t.Fatal("NewAnomalySignal returned nil")
	}
	if sig.ID == "" {
		t.Error("ID is empty — expected a UUID v4 string")
	}
	if sig.DetectedAt.Before(before) || sig.DetectedAt.After(after) {
		t.Errorf("DetectedAt %v is outside [%v, %v]", sig.DetectedAt, before, after)
	}
	if sig.DetectedAt.Location() != time.UTC {
		t.Errorf("DetectedAt location: got %v, want UTC", sig.DetectedAt.Location())
	}
	if sig.Vector != circleai.ThreatVectorBiometricSpoofAttempt {
		t.Errorf("Vector: got %v, want BiometricSpoofAttempt", sig.Vector)
	}
	if sig.AffectedModule != "Circle.AI.Identity" {
		t.Errorf("AffectedModule: got %q, want Circle.AI.Identity", sig.AffectedModule)
	}
	if sig.Description != "liveness check failed" {
		t.Errorf("Description: got %q", sig.Description)
	}
	if v, ok := sig.Evidence["reason"]; !ok || v != "no_blink" {
		t.Errorf("Evidence[reason]: got %q ok=%v, want no_blink", v, ok)
	}
}

func TestNewAnomalySignal_NilEvidenceBecomesEmptyMap(t *testing.T) {
	sig := circleai.NewAnomalySignal(
		circleai.ThreatVectorUnknown,
		0.5,
		"Circle.AI.Companion",
		"unknown anomaly",
		nil,
	)
	if sig == nil {
		t.Fatal("NewAnomalySignal returned nil")
	}
	if sig.Evidence == nil {
		t.Error("Evidence is nil — expected empty map")
	}
	if len(sig.Evidence) != 0 {
		t.Errorf("Evidence len: got %d, want 0", len(sig.Evidence))
	}
}

func TestNewAnomalySignal_UniqueIDs(t *testing.T) {
	const n = 32
	seen := make(map[string]struct{}, n)
	for i := 0; i < n; i++ {
		sig := circleai.NewAnomalySignal(
			circleai.ThreatVectorMemoryAnomaly,
			0.5,
			"Circle.AI.Memory",
			"dup id check",
			nil,
		)
		if sig == nil {
			t.Fatal("NewAnomalySignal returned nil")
		}
		if _, dup := seen[sig.ID]; dup {
			t.Fatalf("duplicate ID generated at iteration %d: %s", i, sig.ID)
		}
		seen[sig.ID] = struct{}{}
	}
}

File: rust/tests/anomaly_signal_test.rs
//! anomaly_signal_test.rs
//!
//! Cross-language confidence-clamp vectors for AnomalySignal::create.
//! Mirrors `fixtures/anomaly_signal_schema.json` `clamp_vectors`.

use std::collections::HashMap;

use circle_ai::security::{AnomalySignal, ThreatVector};

const EPSILON: f32 = 1e-5_f32;

fn approx_eq(a: f32, b: f32) -> bool {
    (a - b).abs() < EPSILON
}

// ─────────────────────────────────────────────────────────────────────────────
// Clamp vectors — must match fixtures/anomaly_signal_schema.json
// ─────────────────────────────────────────────────────────────────────────────

#[test]
fn clamp_above_max() {
    let s = AnomalySignal::create(
        ThreatVector::MemoryAnomaly,
        1.5,
        "Circle.AI.Test",
        "above_max",
        None,
    );
    assert!(
        approx_eq(s.confidence, 1.0),
        "above_max: got {}, expected 1.0",
        s.confidence
    );
}

#[test]
fn clamp_below_min() {
    let s = AnomalySignal::create(
        ThreatVector::MemoryAnomaly,
        -0.3,
        "Circle.AI.Test",
        "below_min",
        None,
    );
    assert!(
        approx_eq(s.confidence, 0.0),
        "below_min: got {}, expected 0.0",
        s.confidence
    );
}

#[test]
fn clamp_at_max() {
    let s = AnomalySignal::create(
        ThreatVector::MemoryAnomaly,
        1.0,
        "Circle.AI.Test",
        "at_max",
        None,
    );
    assert!(
        approx_eq(s.confidence, 1.0),
        "at_max: got {}, expected 1.0",
        s.confidence
    );
}

#[test]
fn clamp_at_min() {
    let s = AnomalySignal::create(
        ThreatVector::MemoryAnomaly,
        0.0,
        "Circle.AI.Test",
        "at_min",
        None,
    );
    assert!(
        approx_eq(s.confidence, 0.0),
        "at_min: got {}, expected 0.0",
        s.confidence
    );
}

#[test]
fn clamp_nominal() {
    let s = AnomalySignal::create(
        ThreatVector::MemoryAnomaly,
        0.7,
        "Circle.AI.Test",
        "nominal",
        None,
    );
    assert!(
        approx_eq(s.confidence, 0.7),
        "nominal: got {}, expected 0.7",
        s.confidence
    );
}

// ─────────────────────────────────────────────────────────────────────────────
// Factory-contract checks
// ─────────────────────────────────────────────────────────────────────────────

#[test]
fn create_stamps_unique_id_and_timestamp() {
    let a = AnomalySignal::create(
        ThreatVector::NetworkPivot,
        0.5,
        "Circle.AI.Test",
        "unique id",
        None,
    );
    let b = AnomalySignal::create(
        ThreatVector::NetworkPivot,
        0.5,
        "Circle.AI.Test",
        "unique id",
        None,
    );
    assert_ne!(a.id, b.id, "factory must stamp a fresh UUID per signal");
    assert!(b.detected_at >= a.detected_at, "timestamps monotonic in test");
}

#[test]
fn evidence_defaults_to_empty_when_none() {
    let s = AnomalySignal::create(
        ThreatVector::StateCorruption,
        0.5,
        "Circle.AI.Test",
        "no evidence",
        None,
    );
    assert!(s.evidence.is_empty());
}

#[test]
fn evidence_round_trips_when_provided() {
    let mut ev = HashMap::new();
    ev.insert("hash".to_string(), "deadbeef".to_string());
    ev.insert("module".to_string(), "biometric".to_string());

    let s = AnomalySignal::create(
        ThreatVector::BiometricSpoofAttempt,
        0.9,
        "Circle.AI.Identity",
        "spoof attempt",
        Some(ev),
    );
    assert_eq!(s.evidence.len(), 2);
    assert_eq!(s.evidence.get("hash").map(String::as_str), Some("deadbeef"));
    assert_eq!(s.evidence.get("module").map(String::as_str), Some("biometric"));
}

#[test]
fn threat_vector_ordinals_match_fixture() {
    // Stable across language ports — see fixtures/anomaly_signal_schema.json.
    assert_eq!(ThreatVector::MemoryAnomaly as u8, 0);
    assert_eq!(ThreatVector::ControlFlowDrift as u8, 1);
    assert_eq!(ThreatVector::PrivilegeEscalation as u8, 2);
    assert_eq!(ThreatVector::BiometricSpoofAttempt as u8, 3);
    assert_eq!(ThreatVector::NetworkPivot as u8, 4);
    assert_eq!(ThreatVector::StateCorruption as u8, 5);
    assert_eq!(ThreatVector::AgentPatchRejected as u8, 6);
    assert_eq!(ThreatVector::Unknown as u8, 7);
}
